### ISSUE TITLE
Feature request - limit number of lines for an issue to 1

### ISSUE THREAD

**Issue by pinkasey** [detekt/detekt#2699]
## Expected Behavior
When a rule finds a warning, the warning never spans more than one line.
If the rule "misbehaves" and reports `entity` that spans multiple lines -
detekt will narrow the issue to a sub-entity that spans 1 line.

I would put this behaviour under a configuration in detekt-config.yml, something like:
`limitIssueToOneLine: true`

## Current Behavior
Some rules mark entire files, classes, or methods as warning - and you can hardly see in IntelliJ that there are additional warnings on the file/class/method.

## Context
This is mostly relevant to the IntelliJ plugin -
There are some rules that mark an entire file/class/method as the issue `entity`.
e.g `MatchingDeclarationName`
This makes it impossible to continue working without addressing the warning somehow
(And what if it's a file I'm just visiting, and wasn't planning on modifying? It's all yellow and hard to read).

Example of what I want to prevent:
![image](https://user-images.githubusercontent.com/1780369/82149843-fe100800-985f-11ea-8ebb-b3e8b4570fe7.png)

**Comment by pinkasey** [comment]
Some additional thoughts:
```
  entity = when (entity) {
    isFile() -> entity.narrowToFirstLine()
    isClass() -> entity.narrowToClassName()
    isCallExpression() -> entity.narrowToCalledMethodName()
    else -> entity.narrowToFirstWord()
  }
```

**Comment by pinkasey** [comment]
Also -
I might find time to contribute here (can't promise)

If there's a preliminary PR - I can finish it
or I might do the entire PR myself

**Comment by arturbosch** [comment]
Thanks for the proposal. I tried it and it works great. I will later post a PR for 1.10.0.

**Comment by BraisGabin** [comment]
But would this apply for detekt or for the intellij plugin?

**Comment by arturbosch** [comment]
I first tried to only implement it at the intellij plugin.
However due to dependency conflict between compiler and embeddable-compiler this must be implemented in detekt or a library.

My second thought was to introduce a `detekt-tooling`/`detekt-embeddable` module which can be consumed by the intellij und sonar plugins with these kind of helpers.

Now I think it makes sense to always report at the identifier not only for classes/objects but also for methods. File related issues can be reported at the package or the first declaration.

I've come up with just extending the `Entity` class:
```kotlin
        /**
         * Factory function which retrieves all needed information from the PsiElement itself.
         */
        fun from(element: PsiElement, offset: Int = 0): Entity {
            val location = Location.from(element, offset)
            return from(element, location)
        }

        /**
         * Create an entity at the location of the identifier of given named declaration.
         */
        fun atName(element: KtNamedDeclaration): Entity = from(element.nameIdentifier ?: element)

        /**
         * Create an entity at the location of the package, first import or first declaration.
         */
        fun atPackageOrFirstDecl(file: KtFile): Entity = from(
            file.packageDirective ?: file.firstChild ?: file
        )
```

**Comment by pinkasey** [comment]
@arturbosch 
That's not exactly what I meant, but thanks for fixing all the rules.

What I wanted was for detekt to **enforce** one line per issue,
so that if some rule makes a mistake of reporting an issue that spans many lines - there's a way to force it to be one line nevertheless, without depending on the author to fix the rule
(detekt/detekt is very fast to respond, but what about other repositories?).

So I was expecting a change in `Detektor.kt`, I think,
Something like this:
```
        fun executeRules(rules: List<BaseRule>) {
            for (rule in rules) {
                rule.visitFile(file, bindingContext)
                for (findingOrig in rule.findings) {
                    val finding = (config.isNarrowEntityFlagOn()) ? finding.narrowEntity() : finding
                      ...
                }
        }

fun Finding.narrowEntity(): Finding {
    return when (val ktElement = this.entity.ktElement){
        is KtFile -> this.copy( entity = Entity.atPackageOrFirstDecl(ktElement))
        is KtFunction -> this.copy( entity = Entity.atName(ktElement))
        else -> this
    }
}
```

**Comment by BraisGabin** [comment]
>@arturbosch 
>That's not exactly what I meant, but thanks for fixing all the rules.
>
>What I wanted was for detekt to **enforce** one line per issue,
so that if some rule makes a mistake of reporting an issue that spans many lines - there's a way to force it to be one line nevertheless, without depending on the author to fix the rule
(detekt/detekt is very fast to respond, but what about other repositories?).

I don't get why a two line error should be reported just like one line error. For example, in the html report there's no problem with that type of errors. It helps the readibility of the snippet if it marks all the problem. I

I think that the @arturbosch's fix here is inside the scope of detekt because it's true that mark ALL the file as an error because a file has too much functions is overkill. But I don't see the reason to make this for all the rules.

Example: 
<img width="402" alt="Captura de pantalla 2020-05-18 a las 19 00 29" src="https://user-images.githubusercontent.com/721244/82239942-dbeab880-9939-11ea-9b5b-0c7d8e42a1fd.png">

Maybe, if this feature have sense for the Intellij plugin (I don't use it), it should be fixed there.

**Comment by arturbosch** [comment]
> @arturbosch
> That's not exactly what I meant, but thanks for fixing all the rules.
> 
> What I wanted was for detekt to **enforce** one line per issue,
> so that if some rule makes a mistake of reporting an issue that spans many lines - there's a way to force it to be one line nevertheless, without depending on the author to fix the rule
> (detekt/detekt is very fast to respond, but what about other repositories?).
> 
> So I was expecting a change in `Detektor.kt`, I think,
> Something like this:
> 
> ```
>         fun executeRules(rules: List<BaseRule>) {
>             for (rule in rules) {
>                 rule.visitFile(file, bindingContext)
>                 for (findingOrig in rule.findings) {
>                     val finding = (config.isNarrowEntityFlagOn()) ? finding.narrowEntity() : finding
>                       ...
>                 }
>         }
> 
> fun Finding.narrowEntity(): Finding {
>     return when (val ktElement = this.entity.ktElement){
>         is KtFile -> this.copy( entity = Entity.atPackageOrFirstDecl(ktElement))
>         is KtFunction -> this.copy( entity = Entity.atName(ktElement))
>         is KtExpression -> this.copy( entity = Entity.atFirstChild(ktElement))
>         else -> this
>     }
> }
> ```

I've tried this approach.

It depends for a KtFile: `is KtFile -> this.copy( entity = Entity.atPackageOrFirstDecl(ktElement))`.
Some rules report the KtFile (which is needed for suppressing) but calculate the Entity by themselves e.g. all formatting rules and some of detekt-rules. We would report them at the wrong location.

`is KtFunction -> this.copy( entity = Entity.atName(ktElement))` is a good rule of thumb which we could implement in a tooling module or something. Unfortunately changing the reported `Finding`s inside the IntelliJ plugin is not possible due to conflicting compiler dependencies (shadowed intellij open api in the kotlin compiler).

I'm on the side with @BraisGabin that we shouldn't narrow the TextRange to a single line for a KtExpression.

> 
> The above code would not compile -
> `Finding.copy` does not exist, for instance. But I hope it conveys my meaning.
> I hope I'm not being rude for bringing this up.

**Comment by pinkasey** [comment]
OK
Thanks!
